grudge/op.py

```diff
@@ -48,37 +84,74 @@
 
 
 from numbers import Number
-from pytools import memoize_on_first_arg
-
-import numpy as np  # noqa
+from functools import reduce
+
+from arraycontext import (
+    ArrayContext,
+    FirstAxisIsElementsTag,
+    make_loopy_program,
+    freeze
+)
+
+from grudge.discretization import DiscretizationCollection
+
+from pytools import (
+    memoize_in,
+    memoize_on_first_arg,
+    keyed_memoize_in
+)
 from pytools.obj_array import obj_array_vectorize, make_obj_array
-import pyopencl.array as cla  # noqa
-from grudge import sym, bind
 
+from meshmode.dof_array import DOFArray
+
+import numpy as np
 import grudge.dof_desc as dof_desc
 
-from meshmode.mesh import BTAG_ALL, BTAG_NONE, BTAG_PARTITION  # noqa
-from meshmode.dof_array import freeze, flatten, unflatten
+from grudge.trace_pair import (  # noqa
+    interior_trace_pair,
+    interior_trace_pairs,
+    connected_ranks,
+    cross_rank_trace_pairs,
+    bdry_trace_pair,
+    bv_trace_pair
+)
 
-from grudge.symbolic.primitives import TracePair
+
+# {{{ Kernel tags
+
+class HasElementwiseMatvecTag(FirstAxisIsElementsTag):
+    """A tag that is applicable to kernel programs indicating that
+    an element-wise matrix product is being performed. This indicates
+    that the first index corresponds to element indices and suggests that
+    the implementation should set element indices as the outermost
+    loop extent.
+    """
+
+# }}}
 
 
-# def interp(dcoll, src, tgt, vec):
+# {{{ Interpolation and projection
+
+# FIXME: Should reintroduce interp and make clear distinctions
+# between projection and interpolations.
+# Related issue: https://github.com/inducer/grudge/issues/38
+# def interp(dcoll: DiscretizationCollection, src, tgt, vec):
 #     from warnings import warn
 #     warn("using 'interp' is deprecated, use 'project' instead.",
-#             DeprecationWarning, stacklevel=2)
+#          DeprecationWarning, stacklevel=2)
 #
-#     return dcoll.project(src, tgt, vec)
+#     return project(dcoll, src, tgt, vec)
 
 
-def project(dcoll, src, tgt, vec):
+def project(dcoll: DiscretizationCollection, src, tgt, vec):
     """Project from one discretization to another, e.g. from the
     volume to the boundary, or from the base to the an overintegrated
     quadrature discretization.
 
-    :arg src: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one
-    :arg tgt: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one
-    :arg vec: a :class:`~meshmode.dof_array.DOFArray`
+    :arg src: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
```

Comment:
Might be worth introducing a type `DOFDescLike = Union[DOFDesc, Hashable]` or something like that? To avoid writing this everywhere.

Reply:
I see what you mean. Though to be honest, I'd prefer to keep this explicit since, right now. It's technically not any hashable that works, only a very specific set of strings. In the future, we may want to consider being a bit more strict about what `dd` and others should be. For now, I'd prefer to leave this as is and just keep the docstring explicit.